File: gin/internal/model/tournaments/game.go
package model_tournaments

import (
	"time"

	"gorm.io/gorm"
	"fmt"
)

type GameWinnerSideType string
const (
	GameWinnerSideType_Player1 GameWinnerSideType = "Player1"
	GameWinnerSideType_Player2 GameWinnerSideType = "Player2"
	GameWinnerSideType_Draw GameWinnerSideType = "Draw"
)

type GameEndedByType string
const (
	GameEndedByType_Normal GameEndedByType = "Normal"
	GameEndedByType_Timeout GameEndedByType = "Timeout"
	GameEndedByType_Concession GameEndedByType = "Concession"
	GameEndedByType_DrawOffer GameEndedByType = "DrawOffer"
)

// GameCreateRequest is the POST body.
type GameCreateRequest struct {
	GameNumber int `json:"game_number"`
	WinnerSide *GameWinnerSideType `json:"winner_side"`
	TurnsPlayed *int `json:"turns_played"`
	DurationSeconds *int `json:"duration_seconds"`
	EndedBy *GameEndedByType `json:"ended_by"`
	ReplayUrl *string `json:"replay_url"`
	MatchID uint `json:"match_id"`
	WinnerID *uint `json:"winner_id"`
}

// GameUpdateRequest is the PUT/PATCH body — all fields optional.
type GameUpdateRequest struct {
	GameNumber *int `json:"game_number"`
	WinnerSide *GameWinnerSideType `json:"winner_side"`
	TurnsPlayed *int `json:"turns_played"`
	DurationSeconds *int `json:"duration_seconds"`
	EndedBy *GameEndedByType `json:"ended_by"`
	ReplayUrl *string `json:"replay_url"`
	MatchID *uint `json:"match_id"`
	WinnerID *uint `json:"winner_id"`
}

// GameResponse is the JSON representation returned by the API.
type GameResponse struct {
	ID        uint      `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	GameNumber int `json:"game_number"`
	WinnerSide *GameWinnerSideType `json:"winner_side"`
	TurnsPlayed *int `json:"turns_played"`
	DurationSeconds *int `json:"duration_seconds"`
	EndedBy *GameEndedByType `json:"ended_by"`
	ReplayUrl *string `json:"replay_url"`
	MatchID uint `json:"match_id"`
	WinnerID *uint `json:"winner_id"`
}

type Game struct {
	gorm.Model
	GameNumber int `gorm:"column:game_number;not null"`
	WinnerSide *GameWinnerSideType `gorm:"column:winner_side"`
	TurnsPlayed *int `gorm:"column:turns_played"`
	DurationSeconds *int `gorm:"column:duration_seconds"`
	EndedBy *GameEndedByType `gorm:"column:ended_by"`
	ReplayUrl *string `gorm:"column:replay_url"`
	MatchID uint `gorm:"column:match_id"`
	WinnerID *uint `gorm:"column:winner_id"`
}

func (m *Game) ToResponse() GameResponse {
	return GameResponse{
		ID:        m.ID,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
		GameNumber: m.GameNumber,
		WinnerSide: m.WinnerSide,
		TurnsPlayed: m.TurnsPlayed,
		DurationSeconds: m.DurationSeconds,
		EndedBy: m.EndedBy,
		ReplayUrl: m.ReplayUrl,
		MatchID: m.MatchID,
		WinnerID: m.WinnerID,
	}
}

func (m *Game) ApplyUpdate(req GameUpdateRequest) {
	if req.GameNumber != nil { m.GameNumber = *req.GameNumber }
	if req.WinnerSide != nil { m.WinnerSide = req.WinnerSide }
	if req.TurnsPlayed != nil { m.TurnsPlayed = req.TurnsPlayed }
	if req.DurationSeconds != nil { m.DurationSeconds = req.DurationSeconds }
	if req.EndedBy != nil { m.EndedBy = req.EndedBy }
	if req.ReplayUrl != nil { m.ReplayUrl = req.ReplayUrl }
	if req.MatchID != nil { m.MatchID = *req.MatchID }
	if req.WinnerID != nil { m.WinnerID = req.WinnerID }
}

func (m *Game) RecordWinner(winnerSide string)  error {
	return fmt.Errorf("RecordWinner: not implemented")
}

func (m *Game) DurationMinutes()  (float64, error) {
	return 0.0, fmt.Errorf("DurationMinutes: not implemented")
}

File: ktor/src/main/kotlin/cards_project/tournaments/model/Game.kt
package cards_project.tournaments.model

import kotlinx.serialization.Serializable
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SqlExpressionBuilder.like
import org.jetbrains.exposed.sql.javatime.CurrentDateTime
import org.jetbrains.exposed.sql.javatime.date
import org.jetbrains.exposed.sql.javatime.datetime
import org.jetbrains.exposed.sql.transactions.transaction

import cards_project.tournaments.model.MatchTable
import cards_project.players.model.PlayerTable

enum class GameWinnerSideType {
    PLAYER1, PLAYER2, DRAW
}

enum class GameEndedByType {
    NORMAL, TIMEOUT, CONCESSION, DRAWOFFER
}

object GameTable : IntIdTable("game") {
    val gameNumber = integer("game_number")
    val winnerSide = enumerationByName<GameWinnerSideType>("winner_side", 50).nullable()
    val turnsPlayed = integer("turns_played").nullable()
    val durationSeconds = integer("duration_seconds").nullable()
    val endedBy = enumerationByName<GameEndedByType>("ended_by", 50).nullable()
    val replayUrl = text("replay_url").nullable()
    val matchId = reference("match_id", MatchTable)
    val winnerId = reference("winner_id", PlayerTable).nullable()
    val createdAt = datetime("created_at").defaultExpression(CurrentDateTime)
    val updatedAt = datetime("updated_at").defaultExpression(CurrentDateTime)
}

data class GameRequest(
    val gameNumber: Int,
    val winnerSide: GameWinnerSideType? = null,
    val turnsPlayed: Int? = null,
    val durationSeconds: Int? = null,
    val endedBy: GameEndedByType? = null,
    val replayUrl: String? = null,
    val matchId: Int,
    val winnerId: Int? = null
)

data class GameResponse(
    val id: Int,
    val gameNumber: Int,
    val winnerSide: GameWinnerSideType? = null,
    val turnsPlayed: Int? = null,
    val durationSeconds: Int? = null,
    val endedBy: GameEndedByType? = null,
    val replayUrl: String? = null,
    val matchId: Int,
    val winnerId: Int?,
    val createdAt: String,
    val updatedAt: String
)

fun ResultRow.toGameResponse() = GameResponse(
    id = this[GameTable.id].value,
    gameNumber = this[GameTable.gameNumber],
    winnerSide = this[GameTable.winnerSide],
    turnsPlayed = this[GameTable.turnsPlayed],
    durationSeconds = this[GameTable.durationSeconds],
    endedBy = this[GameTable.endedBy],
    replayUrl = this[GameTable.replayUrl],
    matchId = this[GameTable.matchId].value,
    winnerId = this[GameTable.winnerId]?.value,
    createdAt = this[GameTable.createdAt].toString(),
    updatedAt = this[GameTable.updatedAt].toString()
)

object GameRepository {

    fun list(skip: Int = 0, limit: Int = 100, q: String? = null): List<GameResponse> = transaction {
        GameTable.selectAll().limit(limit).offset(skip.toLong()).map { it.toGameResponse() }
    }

    fun findById(id: Int): GameResponse? = transaction {
        GameTable.selectAll().where { GameTable.id eq id }.singleOrNull()?.toGameResponse()
    }

    fun create(req: GameRequest): GameResponse = transaction {
        val inserted = GameTable.insertAndGetId {
            it[gameNumber] = req.gameNumber
            it[winnerSide] = req.winnerSide
            it[turnsPlayed] = req.turnsPlayed
            it[durationSeconds] = req.durationSeconds
            it[endedBy] = req.endedBy
            it[replayUrl] = req.replayUrl
            it[matchId] = EntityID(req.matchId, MatchTable)
            req.winnerId?.let { v -> it[winnerId] = EntityID(v, PlayerTable) }
        }
        GameTable.selectAll().where { GameTable.id eq inserted }.single().toGameResponse()
    }

}
